src/blog/understanding-ecmascript-part-2.md

```diff
@@ -0,0 +1,342 @@
+
+The `Receiver` parameter is only used in the case of accessor properties in step 8. It's passed as the **this value** when calling the getter function of an accessor property.
+
+`OrdinaryGet` passes the original `Receiver` throughout the recursion, unchanged (step 3 c). Let's find out where the `Receiver` is originally coming from!
+
+Searching for places where `[[Get]]` is called we find an abstract operation `GetValue` which operates on References. Reference is a specification type, consisting of a base value, the reference name and a strict reference flag. In the case of `o2.foo`, the base value is the Object `o2`, the reference name is the String `"foo"` and the strict reference flag is `false`, since the example code is sloppy.
+
+### Side track: References
+
+Side track: A Reference is not a Record, even though it sounds like it could be. It contains three components, which could equally well be expressed as three fixed named values. Here the spec takes a different approach, and defines References as a higher-level data type. This is because of historical reasons.
+
+### Back to `GetValue`
+
+Let's look at how `GetValue` is defined:
+
+> [`GetValue ( V )`](https://tc39.es/ecma262/#sec-getvalue)
+>
+> 1. `ReturnIfAbrupt(V)`.
+> 2. If `Type(V)` is not `Reference`, return `V`.
+> 3. Let `base` be `GetBase(V)`.
+> 4. If `IsUnresolvableReference(V)` is `true`, throw a `ReferenceError` exception.
+> 5. If `IsPropertyReference(V)` is `true`, then
+> a. If `HasPrimitiveBase(V)` is `true`, then
+> i. Assert: In this case, `base` will never be `undefined` or `null`.
+> ii. Set `base` to `! ToObject(base)`.
+> b. Return `? base.[[Get]](GetReferencedName(V), GetThisValue(V))`.
+> 6. Else,
+> a. Assert: `base` is an Environment Record.
+> b. Return `? base.GetBindingValue(GetReferencedName(V), IsStrictReference(V))`
+
+The reference in our example is `o2.foo` which is a property reference. So we take branch 5. We don't take the branch in 5 a, since the base (`o2`) is not a primitive value (a Boolean, String, Symbol, BigInt or Number).
+
+Then we call `[[Get]]` in step 5 b. The `Receiver` we pass is `GetThisValue(V)`. In this case, it's just the basevalue of the Reference:
+
+> [`GetThisValue( V )`](https://tc39.es/ecma262/#sec-getthisvalue)
+>
+> 1. Assert: `IsPropertyReference(V)` is `true`.
+> 2. If `IsSuperReference(V)` is `true`, then
+> a. Return the value of the `thisValue` component of the reference `V`.
+> 3. Return `GetBase(V)`.
+
+For `o2.foo`, we don't take the branch in step 2, since it's not a super reference (such as `super.foo`), but we take step 3 and return the base value of the reference which is `o2`.
+
+Piecing everything together, we find out that we set the `Receiver` to be the base of the original reference, and then we keep it unchanged during the prototype chain walk. Finally, if the property we find is an accessor property, we use the `Receiver` as the **this value** when calling it.
```

Comment:
What happened to explaining how `o2.foo` gets `2511` before switching gears to getters? There should be a section about the proto walk bottoming out and ultimately going through `o.[[GetOwnProperty]]`.

Reply:
Added the following paragraphs here:

Let's look at how this algorithm works when we access `o2.foo`. First we invoke `OrdinaryGet` with `O` being `o2` and `P` being `"foo"`. `O.[[GetOwnProperty]]` returns `undefined`, since `o2` doesn't have an own property called `"foo"`, so we take the if branch in step 3. In step 3.a, we set `parent` to the prototype of `o2` which is `o1`. `parent` is not `null`, so we don't return in step 3.b. In step 3.c, we call the parent's `[[Get]]` method with property key `"foo"`, and return whatever it returns.

The parent (`o1`) is an ordinary object, so its `[[Get]]` method invokes `OrdinaryGet` again, this time with `O` being `o1` and `P` being `"foo"`. In step 2, `O.[[GetOwnProperty]]("foo")` returns the property descriptor associated to `"foo"`. The property descriptor is not `undefined`, so we don't take the if branch in step 3. Next we execute step 4. The property descriptor is a data descriptor, so we return its value, `99`, in step 4, and we're done.

---- but I didn't change the overall structure of the post, so, the "Receiver" side track is still where it was :)

Reply:
(Also simplified the example and changed the number value because I started liking 99 more.)

Reply:
Added this:


Let's look at how this algorithm works when we access `o2.foo`. First we invoke `OrdinaryGet` with `O` being `o2` and `P` being `"foo"`. `O.[[GetOwnProperty]]("foo")` returns `undefined`, since `o2` doesn't have an own property called `"foo"`, so we take the if branch in step 3. In step 3.a, we set `parent` to the prototype of `o2` which is `o1`. `parent` is not `null`, so we don't return in step 3.b. In step 3.c, we call the parent's `[[Get]]` method with property key `"foo"`, and return whatever it returns.

The parent (`o1`) is an ordinary object, so its `[[Get]]` method invokes `OrdinaryGet` again, this time with `O` being `o1` and `P` being `"foo"`. `o1` has a property called `"foo"`, so in step 2, `O.[[GetOwnProperty]]("foo")` returns the associated Property Descriptor.

[Property Descriptor](https://tc39.es/ecma262/#sec-property-descriptor-specification-type) is a specification type. Data Property Descriptors store the value of the property directly in the `[[Value]]` field. Accessor Property Descriptors store the accessor functions in fields `[[Get]]` and/or `[[Set]]`.

The Property Descriptor returned by `O.[[GetOwnProperty]]("foo")` in step 2 is not `undefined`, so we don't take the if branch in step 3. Next we execute step 4. The Property Descriptor is a Data Descriptor, so we return its `[[Value]]` field, `99`, in step 4, and we're done.

Reply:
(Also changed to 99 because it's a nicer number.)